[contracts] Add docs generator for the contracts API to the `#[define_env]` macro

agryaznov opened the pull request:
Follow-up to #11888

This PR makes the documentation for all the defined host functions available trough the rustdoc of the pallet_contracts.

To build the docs:
- pass `doc` attribute to the macro like [`#[define_env(doc)]`](https://github.com/paritytech/substrate/blob/f8def8c6392fd2014a7afc3bd233d05862fdb61f/frame/contracts/src/wasm/runtime.rs#L999),
  it will make it expand additional `pallet_contracts::api_doc::seal0`, `pallet_contracts::api_doc::seal1`, `...` modules each having its `Api` trait containing functions holding documentation for every host function defined by the macro;
- just run `cargo doc` as usual;

athei commented:
Good approach 🎉. I just have two comments.

First, is there a reason why you made this opt-on with an attribute? Do you think there are cases where we want to skip the docs? You are maybe thinking about the internal `gas` function. But I argue even this should show up because it is still interesting for people who inspect instrumented contracts. The docs should just make it clear that it is internal. I argue it is most important that these docs are complete. Making it opt-in makes it possible that we forget to add the attribute. If there is nothing I am overlooking I am arguing to remove the attribute and just emit the docs for every function all the time.

Second, I think it is not optimal that we need to document private items to see these docs. This adds a lot of noise to the documentation that makes it hard to read if we enable this for docs.rs builds. The API surface is crafted in a way that it only exports types that are relevant to users (not developers) of this pallet. Contract(language) authors for which these docs are meant are part of the user group. What do you think about exporting the various modules emitted here in the crate root?

**lib.rs**
```rust
/// Contains the documentation of the API available to contracts.
///
/// This is not meant to be used by code. It is meant to be consumed by humans through rustdoc.
mod api_doc {
 /// The mod seal_*  go here
}
```

The `mod api_doc` would be emitted by your macro. It is re-exported to the crate root.

athei commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -579,18 +635,33 @@ fn expand_functions(
 pub fn define_env(attr: TokenStream, item: TokenStream) -> TokenStream {
-	if !attr.is_empty() {
-		let msg = "Invalid `define_env` attribute macro: expected no attributes: `#[define_env]`.";
+	if !attr.is_empty() && !(attr.to_string() == "doc".to_string()) {
```
I think we should rename `doc` to `emit_doc`. I think this makes it clearer what is happening. I won't die on that hill if you disagree, though.

agryaznov replied:
to me, concise variant is more elegant and could hardly be misinterpreted

athei replied:
But isn't it the same attribute name used for the doc comments itself?

athei replied:
But isn't it the same attribute name used for rust doc comments itself? To be that seems confusing. Also, why doesn't this PR add this attribute?

agryaznov replied:
> But isn't it the same attribute name used for rust doc comments itself? 

Yeah, that's exactly the point. The idea is to make it short, appropriate and familiar to the user: at the end of the day, it is for generating the rustdoc. I don't see it is confusing, as here `doc` attr is used as input to `define_env(doc)` rather than solo (as when it's for a single doc comment) 

> Also, why doesn't this PR add this attribute to the define_env invocation in runtime.rs?

I was thinking to leave it off by default. But could be done vice versa, why not. Added.

athei replied:
> I don't see it is confusing, as here doc attr is used as input to define_env(doc) rather than solo (as when it's for a single doc comment)

Fair enough.

> I was thinking to leave it off by default. But could be done vice versa, why not. Added.

Most important point for me is that the standard docs.rs builds of this crate contain this documentation. Therefore it should be on.

agryaznov replied:
> Most important point for me is that the standard docs.rs builds of this crate contain this documentation. Therefore it should be on.

Gotcha.  The only objection comes to my mind is that it probably should be turned off for production use not to bloat runtime size in vain. Should we add a note to the macro docs at least?

athei replied:
It should not bloat the runtime size as it is dead code and hence removed by the compiler. At least when using lto.

But you have a point there. I think you should just add `#[cfg(doc)]]` to the emitted module. Then this code will be compiled out for every target but rustdoc.

athei reviewed (commented):
Good approach 🎉

These are mainly nits concerning the emitted handwritten docs. They need to be especially nice since this is how we expect people to learn about the API.

One other thing:

I feel it is not optimal that we need to document private items to see these docs. This adds a lot of noise to the documentation that makes it hard to read if we enable this for docs.rs builds. The API surface is crafted in a way that it only exports types that are relevant to users (not developers) of this pallet. Contract(language) authors for which these docs are meant are part of the user group. What do you think about exporting the various modules emitted here in the crate root?

```rust
/// Contains the documentation of the API available to contracts.
///
/// This is not meant to be used by code. It is meant to be consumed by humans through rustdoc.
///
/// Every function described in this module's sub module's traits uses this sub module's identifier
/// as its imported module name. The identifier of the function is the function's imported name.
/// According to the [wasm spec of imports](https://webassembly.github.io/spec/core/text/modules.html#text-import).		
mod api_doc {
 // The mod seal_*  go here
}
```

The `mod api_doc` would be emitted by your macro. It is re-exported to the crate root.

agryaznov commented:
@athei here goes the next version:

- [x] addressed your feedback,
  in particular, API docs are now expanded to the `pallet_contracts::api_doc` module
- [x] alias functions get automatically explicitly indicated in the docs, 
  with the reference to the origin function instead of duplicated docs
- [x] warning notice for `unstable` functions is now being automatically added to their docs
- [x] tided up some host functions docs along the way,
   seeing them finally in rustdoc for the first time unveiled some minor flaws

athei commented on frame/contracts/src/lib.rs:
```diff
@@ -92,6 +92,7 @@ mod schedule;
 mod storage;
 mod wasm;
 
+pub use wasm::api_doc;
```
Please group with the rest of the public exports further down and not with the module declarations.

athei commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -383,16 +392,111 @@ fn is_valid_special_arg(idx: usize, arg: &FnArg) -> bool {
 	quote! {
 		pub struct Env;
 		#impls
+		pub use docs as api_doc;
```
Why re-export an already public module?

agryaznov replied:
indeed

athei reviewed (commented):
Almost done. Just two tiny things.

xermicus reviewed (approved):
Nice!

agryaznov commented:
bot merge
